ci/images/plugin-compiler/data/build.sh

```diff
@@ -4,74 +4,63 @@ set -xe
+if [ -z "$plugin_name" ]; then
+    usage
+    exit 1
+fi
+
+# if GOOS and GOARCH is not set from docker env, derive it from the host
+# golang-X image.
 if [[ $GOOS == "" ]] && [[ $GOARCH == "" ]]; then
   GOOS=$(go env GOOS)
   GOARCH=$(go env GOARCH)
 fi
 
-if [ -z "$plugin_name" ]; then
-    usage
-    exit 1
-fi
 
 # if arch and os present then update the name of file with those params
 if [[ $GOOS != "" ]] && [[ $GOARCH != "" ]]; then
   plugin_name="${plugin_name%.*}_${CURRENTVERS}_${GOOS}_${GOARCH}.so"
 fi
 
-if [[ $GOOS != "linux" ]];then
-    CGOENABLED=1
-fi
+echo "Building plugin: $plugin_name"
 
-mkdir -p $PLUGIN_BUILD_PATH
-# Plugin's vendor folder, has precedence over the cached vendor'd dependencies from tyk
-yes | cp -r $PLUGIN_SOURCE_PATH/* $PLUGIN_BUILD_PATH || true
+cd $TYK_GW_PATH
+go mod download
+go list -m -f '{{ if not .Main }}{{ .Path }} {{ .Version }}{{ end }}' all > dependencies.txt
 
-cd $PLUGIN_BUILD_PATH
+cd $PLUGIN_SOURCE_PATH
+go get
 
-# Handle if plugin has own vendor folder, and ignore error if not
-[ -f go.mod ] && [ ! -d ./vendor ] && GO111MODULE=on go mod vendor
-# Ensure that go modules not used
-rm -rf go.mod
+# Get plugin dependencies
+go list -m -f '{{ if not .Main }}{{ .Path }} {{ .Version }}{{ end }}' all > dependencies.txt
 
-# We do not need to care which version of Tyk vendored in plugin, since we going to use version inside compiler
-rm -rf $PLUGIN_BUILD_PATH/vendor/github.com/TykTechnologies/tyk
+# for any shared dependency, pin the version to Tyk gateway version
+awk 'NR==FNR{seen[$1]=$2; next} seen[$1] && seen[$1] != $2' $PLUGIN_SOURCE_PATH/dependencies.txt $TYK_GW_PATH/dependencies.txt | while read PKG VER; do
+  go mod edit -replace $PKG=$PKG@$VER
+done
 
-# Copy plugin vendored pkgs to GOPATH
-yes | cp -rf $PLUGIN_BUILD_PATH/vendor/* $GOPATH/src || true \
-        && rm -rf $PLUGIN_BUILD_PATH/vendor
+go mod edit -replace github.com/TykTechnologies/tyk=$TYK_GW_PATH
 
-# Ensure that GW package versions have priorities
+# Run a final go mod tidy to make sure that the go.sum entries are correct for
+# all replaced dependencies.
+go mod tidy
 
-# We can't just copy Tyk dependencies on top of plugin dependencies, since different package versions have different file structures
-# First we need to find which deps GW already has, remove this folders, and after copy fresh versions from GW
-
-# github.com and rest of packages have different nesting levels, so have to handle it separately
-ls -d $TYK_GW_PATH/vendor/github.com/*/* | sed "s|$TYK_GW_PATH/vendor|$GOPATH/src|g" | xargs -d '\n' rm -rf
-ls -d $TYK_GW_PATH/vendor/*/* | sed "s|$TYK_GW_PATH/vendor|$GOPATH/src|g" | grep -v github | xargs -d '\n' rm -rf
-
-# Copy GW dependencies
-yes | cp -rf $TYK_GW_PATH/vendor/* $GOPATH/src
-rm -rf $TYK_GW_PATH/vendor
+CC=$(go env CC)
+# set appropriate X-build gcc binary for arm64.
+if [[ $GOARCH == "arm64" ]] && [[ $GOOS == "linux" ]] ; then
+    CC=aarch64-linux-gnu-gcc
```

Comment:
> When cross-compiling, you must specify a C cross-compiler for cgo to use. You can do this by setting the generic CC_FOR_TARGET or the more specific CC_FOR_${GOOS}_${GOARCH} (for example, CC_FOR_linux_arm) environment variable when building the toolchain using make.bash, or you can set the CC environment variable any time you run the go tool.

As I understand you can declare an unconditional `CC_FOR_linux_arm64` and it's gonna pick it up.

Reply:
That looks neater, will try it out and change accordingly.